### ISSUE TITLE
FFMPEG h.265 -c:v copy workaround

### ISSUE THREAD

**Issue by Zen3515** [blakeblackshear/frigate#7921]
**Describe what you are trying to accomplish and why in non technical terms**

TLDR: ffmpeg can't record video when using -c:v copy on h.265 (hevc) stream on some specific camera.
---

So, I've been in the streaming business for years, and this problem has been plaguing me for a long time.
For some reason, you can't use ffmpeg to record video when using -c:v copy on h.265 (hevc) stream. Further detail [can be seen here](https://superuser.com/questions/1431088/ffmpeg-to-get-an-hevc-rtsp-stream-into-an-mkv-container)

**Describe the solution you'd like**

Ideally, I would want ffmpeg to fix the bug, but I have been waiting for 4 years now, so, I don't think that it'll happen. Next best thing is to add support for the workaround in frigate.

According to [the answer on Stack overflow](https://superuser.com/questions/1431088/ffmpeg-to-get-an-hevc-rtsp-stream-into-an-mkv-container), one needs to use pipe as part of the command for the recording.

```bash
ffmpeg -rtsp_transport tcp -i rtsp://xx -an -vcodec copy -bsf:v hevc_metadata=tick_rate=1 -f hevc - | ffmpeg -f hevc -i - -c copy out.mkv -y
```

But when we set the ffmpeg output to include pipe, [Popen](https://github.com/blakeblackshear/frigate/blob/367d7244d37f4e48ce1760b319c0fe141ab58711/frigate/video.py#L128-L151) doesn't work without `shell=True` or manual pipe between two process.

So, I propose that we allow for `shell=True` when user wish to use it. I've tested it and found that it is working very well.

```python
def start_or_restart_ffmpeg(
    ffmpeg_cmd, logger, logpipe: LogPipe, frame_size=None, ffmpeg_process=None
):
    if ffmpeg_process is not None:
        stop_ffmpeg(ffmpeg_process, logger)

    if frame_size is None:
        process = sp.Popen(
            ' '.join(ffmpeg_cmd),
            shell=True,
            stdout=sp.DEVNULL,
            stderr=logpipe,
            stdin=sp.DEVNULL,
            start_new_session=True,
        )
    else:
        process = sp.Popen(
            ' '.join(ffmpeg_cmd),
            shell=True,
            stdout=sp.PIPE,
            stderr=logpipe,
            stdin=sp.DEVNULL,
            bufsize=frame_size * 10,
            start_new_session=True,
        )
    return process
```

But due to security reasons, I don't think shell should be used by default, but it should be possible to toggle.

**Describe alternatives you've considered**

`shell=True` is simpler to implement, otherwise, you could split the recording command to many processes that are connected via pipe.

Or you could let people know that h.265 recording is possible, but not with `-c:v copy` in the documentation

**Additional context**

Here are some errors to detect this problem
- firgate log
```
2023-09-08 16:41:40.826065605  [2023-09-08 16:41:40] frigate.record                 WARNING : Discarding a corrupt recording segment: /tmp/cache/Entrance-F-20230908164124.mp4
2023-09-08 16:41:50.818924095  [2023-09-08 16:41:50] frigate.record                 WARNING : Failed to probe corrupt segment /tmp/cache/Entrance-F-20230908164137.mp4: 0 - b'[extract_extradata @ 0x56107c937ac0] No start code is found.\n/tmp/cache/Entrance-F-20230908164137.mp4: Invalid data found when processing input\n'
```

Here is how I currently work around this problem.
1. Edit [video.py](https://github.com/blakeblackshear/frigate/blob/367d7244d37f4e48ce1760b319c0fe141ab58711/frigate/video.py#L128-L151) file and mount it into docker with the stated above function modified (`start_or_restart_ffmpeg`)
2. Set the relevant config
```yaml

ffmpeg:
  global_args: -hide_banner -loglevel error
  hwaccel_args:
    - -hwaccel
    - auto
cameras:
  example_camera_1:
    ffmpeg:
      input_args: -user_agent \'FFmpeg Frigate/0.12.1-367d724\' -avoid_negative_ts make_zero -fflags +genpts+discardcorrupt -rtsp_transport tcp -timeout 5000000 -use_wallclock_as_timestamps 1
      output_args:
        record: -c:v copy -c:a copy -f hevc - | /usr/lib/btbn-ffmpeg/bin/ffmpeg -i - -c:v copy -c:a copy -f segment -segment_time 10 -segment_format mp4 -segment_format_options movflags=faststart+frag_keyframe+empty_moov+default_base_moof -reset_timestamps 1 -segment_atclocktime 1 -strftime 1
```

BTW, I've tried complie latest version of ffmpeg, the problem still exist.

**Comment by NickM-27** [comment]
I don't think this is correct, I have multiple cameras set to h.265 and they are recording with `preset-record-generic-audio-aac without issues

> Further detail [can be seen here](https://superuser.com/questions/1431088/ffmpeg-to-get-an-hevc-rtsp-stream-into-an-mkv-container)

that is referring to mkv container, frigate does not use mkv container

**Comment by NickM-27** [comment]
Here is the ffmpeg command frigate is using:

`ffmpeg -hide_banner -loglevel warning -threads 2 -user_agent FFmpeg Frigate/0.13.0-fe800f51 -rtsp_transport tcp -timeout 5000000 -i rtsp://127.0.0.1:8554/front_cam_full -f segment -segment_time 10 -segment_format mp4 -reset_timestamps 1 -strftime 1 -c:v copy -c:a aac /tmp/cache/front_cam-%Y%m%d%H%M%S.mp4`

**Comment by Zen3515** [comment]
> that is referring to mkv container, frigate does not use mkv container

That's not the point, I found this problem on the Chinese ip cameras

> Here is the ffmpeg command frigate is using:
> 
> `ffmpeg -hide_banner -loglevel warning -threads 2 -user_agent FFmpeg Frigate/0.13.0-fe800f51 -rtsp_transport tcp -timeout 5000000 -i rtsp://127.0.0.1:8554/front_cam_full -f segment -segment_time 10 -segment_format mp4 -reset_timestamps 1 -strftime 1 -c:v copy -c:a aac /tmp/cache/front_cam-%Y%m%d%H%M%S.mp4`
> 
> and ffprobe of the stream
> 
> ```
> Input #0, rtsp, from 'rtsp://192.168.50.106:8554/front_cam_full':
>   Metadata:
>     title           : go2rtc/1.7.1
>   Duration: N/A, start: 0.000000, bitrate: N/A
>   Stream #0:0: Video: hevc (Main), yuvj420p(pc), 5120x1440 [SAR 1:1 DAR 32:9], 20 fps, 20 tbr, 90k tbn
>   Stream #0:1: Audio: pcm_mulaw, 8000 Hz, mono, s16, 64 kb/s
> ```

If I use that exact command, it'll not work for my case.  
I don't know exactly who to blame here, but the problem does indeed exist, I have been experiencing this problem for years.

To be precise, I found three camera models having these issues, two from Hikvision, another from some OEM brand.

**Comment by NickM-27** [comment]
the TLDR and post in general presents the issue as though it is the case universally, where as now it is clear this only applies to specific cameras. 

The example I provided above is running the stream from a `Hikvision DS-2CD2387G2P-LSU/SL` so it would seem this doesn't even apply universally to a brand of cameras. 

Just for clarity sake, I am also running h.265 on Amcrest & Dahua cameras without issues.

**Comment by Zen3515** [comment]
> the TLDR and post in general presents the issue as though it is the case universally for h.265, where as now it is clear this only applies to specific cameras.
> 
> The example I provided above is running the stream from a `Hikvision DS-2CD2387G2P-LSU/SL` so it would seem this doesn't even apply universally to a brand of cameras.
> 
> Just for clarity sake, I am also running h.265 on Amcrest & Dahua cameras without issues.

Ok, I fixed the tldr.

Indeed, judging from some random people [ffmpeg commnad](https://github.com/imaginator/system/blob/4f248fe4336559c8cce1b67ce0b9d2b674ad9a49/salt/motion/files/record-video-camera01.service) their camera seems to be Hikvision too, and he has to use the workaround.

And maybe more people have experienced this in the past but have given up on h.265, or just re-encode it. For me, it took me a year before reporting this issue.

**Comment by NickM-27** [comment]
Right, I have provided support for many users here and using h.265 (without issues) is definitely not uncommon, so I wanted to clarify that point.

Out of curiosity, I may try enabling Hikvision's h.265+ smart encoding and see if it causes this or any other issues with the recording.

Does this issue also apply when the problematic ffmpeg command is run on the segments that are saved from these cameras?

**Comment by NickM-27** [comment]
looking into those issues more, this seems to be the problem:

> ffmpeg looks for codec extradata only in first packet but the RTSP input appears to not have it.

which makes sense, it is a bug in the camera software that ffmpeg does not handle. 

I would be curious if you ran your stream through go2rtc 1.7.1 (or some other restreamer like mediamtx) if the issue would be alleviated. There are a couple different things that could be tried in go2rtc that might fix this there.

**Comment by Zen3515** [comment]
> looking into those issues more, this seems to be the problem:
> 
> > ffmpeg looks for codec extradata only in first packet but the RTSP input appears to not have it.
> 
> which makes sense, it is a bug in the camera software that ffmpeg does not handle.
> 
> I would be curious if you ran your stream through go2rtc 1.7.1 (or some other restreamer like mediamtx) if the issue would be alleviated. There are a couple different things that could be tried in go2rtc that might fix this there.

I've been using go2rtc 1.7.1 this whole time but I haven't tried mediamtx yet.

**Comment by NickM-27** [comment]
go2rtc 1.7 was only released a few weeks ago, based on your timeline it would seem you may be on the go2rtc 1.2 that is included in frigate 0.12

In any case, I'd be curious how go2rtc 1.7.1 does directly and using the ffmpeg handler for the stream and then the restream is used for recording. 

Similar for mediamtx if go2rtc doesn't work

**Comment by Zen3515** [comment]
> go2rtc 1.7 was only released a few weeks ago, based on your timeline it would seem you may be on the go2rtc 1.2 that is included in frigate 0.12
> 
> In any case, I'd be curious how go2rtc 1.7.1 does directly and using the ffmpeg handler for the stream and then the restream is used for recording.
> 
> Similar for mediamtx if go2rtc doesn't work

I've used rtsp-simple-server in the past (now it's called mediamtx) when h.265 is not common yet. Juding from my experience, it gives a better shot than go2rtc, since go2rtc rtsp implementation is quite minimal.

My go2rtc is of latest version, the older version suffered from [this](https://github.com/AlexxIT/go2rtc/commit/4baa3f5588e122c73d0ec8cba966199aafe87edf). So, I decided to not use the built-in one. 
![image](https://github.com/blakeblackshear/frigate/assets/7106408/b7b9aa09-bac2-4f5c-81db-86404066af89)

This image shows that if shell=True were to use, it's a viable workaround.

![image](https://github.com/blakeblackshear/frigate/assets/7106408/69312ec9-ac71-42f4-9e4f-ae1570dbe4ce)

To summarize,
- I want to confirm that go2rtc restreaming doesn't improve the situation.
- mediamtx might give a better chance than go2rtc, but I don't want to use it, how many containers do I have to spin up just for recording the cctv. The reason that I don't switch to mediamtx instead is because webrtc implementation of go2rtc, in my opinion, might be better.
- I made this issue to ask if it's possible to include an option to enable shell mode if the user desires it. As a workaround for this issue.

**Comment by NickM-27** [comment]
I think even if go2rtc native rtsp doesn't fix, using the ffmpeg input has a decent chance, because the stream that ffmpeg puts out should follow the protocol and include the media information in the first packet. 

We need to understand the issue before enabling shell for example because it's something that would need to be maintained and documented in a way that isn't confusing.

**Comment by Zen3515** [comment]
> I think even if go2rtc native rtsp doesn't fix, using the ffmpeg input has a decent chance, because the stream that ffmpeg puts out should follow the protocol and include the media information in the first packet.
> 
> We need to understand the issue before enabling shell for example because it's something that would need to be maintained and documented in a way that isn't confusing.

I don't quite follow you, what do you mean by `ffmpeg input`, is it something I can set in frigate, or you want me to chain the stream inside go2rtc?

**Comment by NickM-27** [comment]
So this type of setup would be native go2rtc handling

```yaml
go2rtc:
  streams:
    test:
      - "rtsp://{FRIGATE_RTSP_USER}:{FRIGATE_RTSP_PW}@192.168.50.152/streaming/channels/101"

cameras:
  test:
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/test
           input_args: preset-rtsp-restream
```

this is an example of using the ffmpeg input where ffmpeg handles the input and then go2rtc restreams

```yaml
go2rtc:
  streams:
    test:
      - "ffmpeg:rtsp://{FRIGATE_RTSP_USER}:{FRIGATE_RTSP_PW}@192.168.50.152/streaming/channels/101"

cameras:
  test:
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/test
           input_args: preset-rtsp-restream
```

**Comment by Zen3515** [comment]
For this problem, there's no direct way to fix it.
But I found the perfect workaround, rather than patching frigate to support `shell=true`

I found that go2rtc [can pipe output of the exec command](https://github.com/AlexxIT/go2rtc#source-exec)
> **pipe** reads data from app stdout in different formats: MJPEG, H.264/H.265 bitstream, MPEG-TS.

Taking advantage of this, you could just add this to go2rtc and be fine on whatever program you want to use the stream in the future.

```yaml

streams:
  example:
    - exec:ffmpeg -fflags +genpts+igndts -use_wallclock_as_timestamps 1 -rtsp_transport tcp -i rtsp://exampleuser:examplepass@ipcam-01.example.com:554/Streaming/Channels/101?transportmode=unicast&profile=Profile_1 -c:v copy -c:a copy -f hevc -
    - "isapi://exampleuser:examplepass@ipcam-01.example.com:80/"
```

So this issue could be considered closed now, not just stale and closed.

**Comment by YBonline** [comment]
I acquired a property that had 8x WBOX 0E-40DFIK10 (a Hikvision rebrand from my understanding) pre-installed. I initially switched them to H264 resolving my issue, but data storage and bandwidth made me want to take advantage of H265, and I finally had some time to research this, and Google brought me to this issue. I had identical symptoms to Zen3515 All of them exhibited similar issues. In fact, it was quite confusing since the substream seems to not produce errors, and the main stream plays fine in the live view, but it cannot record the segments with the aforementioned "corrupted segment" error appearing the in the logs.

His solution also seems to fully resolve my issue as well, so I figured I'd mention it'd be nice if this workaround could make it into the documentation of camera specific issues? If not, at least the search engines should hopefully see this camera model number to help others find the solution.

**Comment by yehpo** [comment]
I seem to have the same issue with 2 of my non-standard Hikvision cameras:
- DS-2CD63C5G0-IVS
- DS-2CD2786G2-IZS
I have tried the workaround posted by @Zen3515 above, but I can't seem to get it to work. I have copy pasted the string below, replacing the stream URL with my own, but that doesn't seem to change anything. My relevant is code below. Please someone advise if I am missing something obvious.

```
go2rtc:
  streams:
    Packing_Area_Camera:
      - exec:ffmpeg -fflags +genpts+igndts -use_wallclock_as_timestamps 1 -rtsp_transport tcp -i rtsp://{FRIGATE_RTSP_USER}:{FRIGATE_RTSP_PASSWORD}@192.168.1.48:554/Streaming/channels/101?transportmode=unicast&profile=Profile_1 -c:v copy -c:a copy -f hevc -
      - "isapi://{FRIGATE_RTSP_USER}:{FRIGATE_RTSP_PASSWORD}@192.168.1.48:80/"

cameras:
  Packing_Area_Camera:
    ffmpeg:
      hwaccel_args: preset-vaapi
      output_args:
        record: preset-record-generic-audio-aac
      inputs:
        - path: rtsp://{FRIGATE_RTSP_USER}:{FRIGATE_RTSP_PASSWORD}@192.168.1.48:554/Streaming/channels/102
          roles:
            - detect
        - path: rtsp://{FRIGATE_RTSP_USER}:{FRIGATE_RTSP_PASSWORD}@192.168.1.48:554/Streaming/channels/101
          roles:
            - record
    detect:
      width: 960
      height: 720
      fps: 10
    record:
      enabled: True
```
